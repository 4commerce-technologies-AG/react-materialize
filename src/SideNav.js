import React, { Fragment, useEffect, useRef } from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';

import idgen from './idgen';

const SideNav = ({
  options,
  id,
  className,
  children,
  trigger,
  fixed,
  ...props
}) => {
  const instance = useRef(null);
  const sidenavRef = useRef(null);
  const _id = id || `sidenav_${idgen()}`;
  const classNames = cx(
    'sidenav',
    { 'sidenav-fixed': fixed || !trigger },
    className
  );

  useEffect(() => {
    instance.current = M.Sidenav.init(sidenavRef.current, options);

    return () => {
      if (instance.current) {
        instance.current.destroy();
      }
    };
  }, [options]);

  const renderTrigger = () => {
    if (!trigger) {
      return;
    }
    const triggerView = fixed ? 'hide-on-large-only' : 'show-on-large';
    const classNames = cx(
      trigger.props.className,
      triggerView,
      'sidenav-trigger'
    );
    return React.cloneElement(trigger, {
      ref: t => (t._trigger = `[data-target=${_id}]`),
      'data-target': _id,
      className: classNames
    });
  };

  return (
    <Fragment>
      {renderTrigger()}
      <ul
        ref={el => (sidenavRef.current = el)}
        id={_id}
        className={classNames}
        {...props}
      >
        {children}
      </ul>
    </Fragment>
  );
};

SideNav.propTypes = {
  /**
   * Adds sidenav-fixed class to sidenav
   */
  fixed: PropTypes.bool,
  /**
   * sidenav id. If none is passed, an id will be generated.
   */
  id: PropTypes.string,
  /**
   * Component that is rendered to trigger the sidenav
   */
  trigger: PropTypes.node,
  /**
   * Options hash for the sidenav.
   * More info: http://materializecss.com/side-nav.html#options
   */
  options: PropTypes.shape({
    edge: PropTypes.oneOf(['left', 'right']),
    draggable: PropTypes.bool
  }),
  /**
   * Additional classes added to 'sidenav'
   */
  className: PropTypes.string,
  children: PropTypes.node
};

export default SideNav;
