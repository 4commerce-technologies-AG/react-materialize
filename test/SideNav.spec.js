import React from 'react';
import { shallow, mount } from 'enzyme';
import SideNav from '../src/SideNav';
import mocker from './helper/new-mocker';

describe('<SideNav />', () => {
  let wrapper;
  const sideNavInitMock = jest.fn();
  const sideNavInstanceDestroyMock = jest.fn();
  const sideNavMock = {
    init: (el, options) => {
      sideNavInitMock(options);
      return {
        destroy: sideNavInstanceDestroyMock
      };
    }
  };
  const restore = mocker('Sidenav', sideNavMock);
  const trigger = <button className="trigger">Show sidenav</button>;

  afterAll(() => {
    restore();
  });

  test('renders', () => {
    wrapper = shallow(<SideNav trigger={trigger} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('can be `fixed`', () => {
    wrapper = shallow(<SideNav trigger={trigger} fixed />);
    expect(wrapper).toMatchSnapshot();
  });

  test('is `fixed` if no trigger is passed', () => {
    wrapper = shallow(<SideNav />);
    expect(wrapper).toMatchSnapshot();
  });

  test('renders', () => {
    wrapper = shallow(<SideNav trigger={trigger} id="test" />);
    expect(wrapper).toMatchSnapshot();
  });

  test('renders children', () => {
    wrapper = shallow(
      <SideNav trigger={trigger}>
        <span className="test-child" />
      </SideNav>
    );

    expect(wrapper).toMatchSnapshot();
  });

  test('should call sideNav with the given options', () => {
    const options = {
      closeOnClick: true,
      edge: 'right'
    };
    mount(<SideNav trigger={trigger} options={options} />);

    expect(sideNavInitMock).toHaveBeenCalledWith(options);
  });

  test('should be destroyed when unmounted', () => {
    const component = shallow(<SideNav />);
    component.unmount();

    expect(sideNavInstanceDestroyMock).toHaveBeenCalled();
  });
});
