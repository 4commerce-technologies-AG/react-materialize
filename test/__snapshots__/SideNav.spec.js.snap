// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SideNav /> can be \`fixed\` 1`] = `
<SideNav
  fixed={true}
  trigger={
    <button
      className="trigger"
    >
      Show sidenav
    </button>
  }
>
  <button
    className="trigger hide-on-large-only sidenav-trigger"
    data-target="sidenav_mockId"
  >
    Show sidenav
  </button>
  <ul
    className="sidenav sidenav-fixed"
    id="sidenav_mockId"
  />
</SideNav>
`;

exports[`<SideNav /> is \`fixed\` if no trigger is passed 1`] = `
<SideNav>
  <ul
    className="sidenav sidenav-fixed"
    id="sidenav_mockId"
  />
</SideNav>
`;

exports[`<SideNav /> renders 1`] = `
<SideNav
  trigger={
    <button
      className="trigger"
    >
      Show sidenav
    </button>
  }
>
  <button
    className="trigger show-on-large sidenav-trigger"
    data-target="sidenav_mockId"
  >
    Show sidenav
  </button>
  <ul
    className="sidenav"
    id="sidenav_mockId"
  />
</SideNav>
`;

exports[`<SideNav /> renders 2`] = `
<SideNav
  id="test"
  trigger={
    <button
      className="trigger"
    >
      Show sidenav
    </button>
  }
>
  <button
    className="trigger show-on-large sidenav-trigger"
    data-target="test"
  >
    Show sidenav
  </button>
  <ul
    className="sidenav"
    id="test"
  />
</SideNav>
`;

exports[`<SideNav /> renders children 1`] = `
<SideNav
  trigger={
    <button
      className="trigger"
    >
      Show sidenav
    </button>
  }
>
  <button
    className="trigger show-on-large sidenav-trigger"
    data-target="sidenav_mockId"
  >
    Show sidenav
  </button>
  <ul
    className="sidenav"
    id="sidenav_mockId"
  >
    <span
      className="test-child"
    />
  </ul>
</SideNav>
`;
